File: runtime/host/deploy.ex
defmodule Workbooks.Deploy do
  @moduledoc """
  The deploy-kit — stand up the ONE runtime OCI image, locally or on a cloud
  machine. Radically simpler than the legacy multi-cell kit: one image, two places.

    * `local` — the same image in a local Linux container (krunvm/libkrun on mac),
      managed by a launchd LaunchAgent so it runs on login + survives app quit.
      The desktop "daemon"; the Tauri shell drives it. Cloud-identical isolation
      (bwrap needs Linux namespaces, absent on bare macOS).
    * `cloud` — the same image on one machine (generic; not fly-locked).

  Built for AGENTS as much as humans: every verb is non-interactive, idempotent,
  returns a tagged `{:ok | :error, map}` (rendered human OR `--json`), and exits
  non-zero on failure. `doctor` self-heals prereqs; `verify` proves the LIVE
  runtime answers. Backends sit behind a seam (`Krunvm` today; podman/docker/WSL2
  later). Reached via `wb deploy <verb> [--json]`.
  """
  alias Workbooks.Deploy.{Krunvm, Image, Backend, Config}

  @default_file "deployment.org"

  @template_local """
  # A Workbooks deployment — single-tenant, all-local (runs in a krunvm microVM,
  # cloud-identical isolation). Edit, then:  wb deploy validate  →  wb deploy apply
  * deployment :deployment:
    :PROPERTIES:
    :ENGINE_PLACE: local
    :TENANCY_MODE: single
    :STORAGE:      local-fs
    :DATABASE:     sqlite
    :AUTH:         trusted
    # :PROFILE:    path/to/your/agent/profile
    :END:
  """

  @template_cloud """
  # A Workbooks deployment — multi-tenant SaaS in the cloud. Fill the <...>.
  # SECRETS live in your ENV, never here:  WB_S3_KEY  WB_S3_SECRET  WB_DATABASE_URL
  # Then:  wb deploy validate  →  wb deploy apply
  * deployment :deployment:
    :PROPERTIES:
    :ENGINE_PLACE:     cloud
    :PROVIDER:         fly
    :APP:              <your-app-name>
    :REGION:           sjc
    :TENANCY_MODE:     multi
    :STORAGE:          s3
    :STORAGE_ENDPOINT: https://s3.us-east-1.amazonaws.com
    :STORAGE_BUCKET:   <your-bucket>
    :STORAGE_REGION:   us-east-1
    :DATABASE:         postgres
    :AUTH:             clerk
    :ISSUER:           https://<your-clerk-domain>
    # :PROFILE:        path/to/your/agent/profile
    :END:
  """

  @doc "Scaffold a deployment.org from a preset (`local` | `cloud`). The starting point."
  def init(preset \\ "local", opts \\ []) do
    file = Keyword.get(opts, :file, @default_file)
    force? = Keyword.get(opts, :force, false)

    cond do
      template(preset) == nil ->
        err("unknown preset '#{preset}' — try: local | cloud", %{presets: ["local", "cloud"]})

      File.exists?(file) and not force? ->
        err("#{file} already exists — edit it, or `wb deploy init #{preset} --force` to replace it", %{file: file})

      true ->
        File.write!(file, template(preset))
        ok("wrote #{file} (#{preset}) — edit it, then `wb deploy validate` → `wb deploy apply`", %{file: file, preset: preset})
    end
  end

  defp template("local"), do: @template_local
  defp template("cloud"), do: @template_cloud
  defp template("cloud-saas"), do: @template_cloud
  defp template(_), do: nil

  @doc "Coherence-check a deployment.org without deploying (the write-then-submit gate)."
  def validate(file) do
    with :ok <- exists(file),
         {:ok, p} <- Config.parse(file) do
      case Config.validate(p) do
        :ok -> ok("valid — #{Config.summary(p)}", %{valid: true})
        {:error, issues} -> err("invalid deployment:\n  - " <> Enum.join(issues, "\n  - "), %{valid: false, issues: issues})
      end
    else
      {:error, msg} -> err(msg, %{})
    end
  end

  defp exists(file) do
    if File.exists?(file), do: :ok, else: {:error, "no #{file} — run `wb deploy init` to scaffold one"}
  end

  @doc """
  Apply a `deployment.org`: validate, then converge to it. ENGINE_PLACE picks the
  target (local krunvm vs the cloud provider); TENANCY_MODE + the BYOD STORAGE/
  DATABASE axes + PROFILE flow to the engine as env. Idempotent.
  """
  def apply(file) do
    with :ok <- exists(file),
         {:ok, p} <- Config.parse(file),
         :ok <- coherent(p) do
      env = Config.to_env(p)

      case Config.place(p) do
        "local" -> local(env: env)
        "cloud" -> cloud_apply(p, env)
        other -> err("ENGINE_PLACE must be local|cloud (got #{inspect(other)})", %{})
      end
    else
      {:error, issues} when is_list(issues) -> err("invalid deployment:\n  - " <> Enum.join(issues, "\n  - "), %{issues: issues})
      {:error, msg} -> err(msg, %{})
    end
  end

  defp coherent(p) do
    case Config.validate(p), do: (:ok -> :ok; {:error, issues} -> {:error, issues})
  end

  # Cloud is one concept; the provider is config (PROVIDER, default fly) — the
  # extension point, not a user-facing menu. The config env becomes the engine's.
  defp cloud_apply(p, env) do
    provider = Config.provider(p)

    case Backend.resolve(provider) do
      {:provider, _pl, boot} ->
        case Backend.run_provider(boot, "up", Map.put(env, "WB_IMAGE", Image.ref())) do
          {:ok, out} -> ok("cloud deploy (#{provider}) — #{Config.summary(p)}:\n#{String.trim_trailing(out)}", %{provider: provider})
          {:error, out} -> err("cloud deploy (#{provider}) failed:\n#{String.trim_trailing(out)}", %{provider: provider})
        end

      _ ->
        err("cloud PROVIDER=#{provider} has no recipe at deploy/providers/#{provider}", %{provider: provider})
    end
  end

  @doc "Check + self-heal prerequisites (idempotent). The first thing an agent runs."
  def doctor do
    case Krunvm.preflight() do
      :ok ->
        ok("prereqs OK", %{krunvm: true, apfs_volume: true})

      {:error, :apfs_volume_missing, _} ->
        case Krunvm.ensure_apfs_volume() do
          :ok -> ok("created the case-sensitive APFS volume; prereqs OK now", %{krunvm: true, apfs_volume: :created})
          {:error, m} -> err(m, %{krunvm: true, apfs_volume: false})
        end

      {:error, reason, msg} ->
        err(msg, %{reason: reason})
    end
  end

  @doc "Bring up the local containerized runtime daemon. Idempotent (converges)."
  def local(opts \\ []) do
    image = Keyword.get(opts, :image, Image.ref())
    env = Keyword.get(opts, :env, %{})
    host_port = Keyword.get(opts, :host_port, free_host_port())

    with {:ok, _, _} <- doctor(),
         {:ok, info} <- Krunvm.create(image, host_port: host_port),
         {:ok, _} <- Krunvm.install_agent(env) do
      base = %{url: info.url, image: image, data_dir: info.data_dir, agent: Krunvm.label()}

      case await_discovery(15_000) do
        {:ok, disc} ->
          ok("local runtime up — #{info.url} (runs on login, survives app quit; `wb deploy down` to stop)",
             Map.merge(base, %{state: "up", token_preview: String.slice(disc["token"] || "", 0, 8)}))

        {:error, :timeout} ->
          ok("VM + agent installed (#{info.url}) but no discovery yet — `wb deploy logs`/`verify`. Image may still be booting.",
             Map.merge(base, %{state: "starting"}))
      end
    else
      {:error, msg, data} -> err(msg, data)
      {:error, msg} -> err(msg, %{})
    end
  end

  @doc "Daemon state — local by default, or a cloud deployment's state given its config file."
  def status(file \\ nil), do: lifecycle(file, &local_status/0, "status")

  defp local_status do
    vm? = Krunvm.exists?()

    {runtime, fields} =
      case Krunvm.discovery() do
        {:ok, d} -> {"up — http://127.0.0.1:#{d["port"]} (pid #{d["pid"]})", %{state: "up", port: d["port"], pid: d["pid"]}}
        _ -> {"no discovery file", %{state: "down"}}
      end

    ok("local runtime: microVM #{if(vm?, do: "present", else: "absent")}; runtime #{runtime}",
       Map.merge(fields, %{microvm: vm?, agent: Krunvm.label()}))
  end

  @doc "Prove the LIVE runtime answers — local daemon, or a cloud deployment's URL."
  def verify(file \\ nil) do
    case place_of(file) do
      {:cloud, p} -> cloud_verify(p)
      _ -> local_verify()
    end
  end

  defp local_verify do
    with {:ok, d} <- Krunvm.discovery(),
         port when is_integer(port) <- d["port"],
         {:ok, status, body} <- http_get("http://127.0.0.1:#{port}/health", d["token"]) do
      if status == 200 and String.contains?(body, "ok"),
        do: ok("runtime healthy — http://127.0.0.1:#{port}/health → 200", %{state: "healthy", port: port}),
        else: err("runtime reachable but unhealthy (HTTP #{status})", %{state: "unhealthy", port: port, http: status})
    else
      {:error, :no_discovery} -> err("no discovery file — is the daemon up? `wb deploy local`", %{state: "down"})
      {:error, reason} -> err("runtime not reachable: #{inspect(reason)}", %{state: "unreachable"})
      _ -> err("runtime not reachable", %{state: "unreachable"})
    end
  end

  @doc "Tear down — local daemon by default, or a cloud deployment given its config file."
  def down(file \\ nil), do: lifecycle(file, &local_down/0, "down")

  defp local_down do
    Krunvm.down()
    ok("local runtime down (data + APFS volume preserved)", %{state: "down"})
  end

  @doc "Where logs are — local daemon, or a cloud deployment's logs given its config file."
  def logs(file \\ nil), do: lifecycle(file, &local_logs/0, "logs")

  defp local_logs do
    dir = Path.join([System.user_home!(), "Library", "Application Support", "sh.workbooks", "logs"])
    out = Path.join(dir, "runtime.out.log")
    err_log = Path.join(dir, "runtime.err.log")
    ok("tail -f #{err_log} #{out}", %{stdout: out, stderr: err_log})
  end

  # ---- local|cloud dispatch for the lifecycle verbs --------------------------
  # No file → local. A file → its ENGINE_PLACE decides (cloud runs the provider).
  defp lifecycle(file, local_fun, action) do
    case place_of(file) do
      {:cloud, p} -> cloud_action(p, action)
      {:error, msg} -> err(msg, %{})
      _ -> local_fun.()
    end
  end

  defp place_of(nil), do: :local

  defp place_of(file) do
    case Config.parse(file) do
      {:ok, p} -> if Config.place(p) == "cloud", do: {:cloud, p}, else: {:local, p}
      {:error, msg} -> {:error, msg}
    end
  end

  defp cloud_action(p, action) do
    provider = Config.provider(p)

    case Backend.resolve(provider) do
      {:provider, _pl, boot} ->
        case Backend.run_provider(boot, action, Map.put(Config.to_env(p), "WB_IMAGE", Image.ref())) do
          {:ok, out} -> ok("cloud #{action} (#{provider}):\n#{String.trim_trailing(out)}", %{provider: provider, action: action})
          {:error, out} -> err("cloud #{action} (#{provider}) failed:\n#{String.trim_trailing(out)}", %{provider: provider, action: action})
        end

      _ ->
        err("cloud PROVIDER=#{provider} has no recipe at deploy/providers/#{provider}", %{provider: provider})
    end
  end

  # Cloud liveness: ask the provider for the public URL, then GET /health.
  defp cloud_verify(p) do
    provider = Config.provider(p)

    with {:provider, _pl, boot} <- Backend.resolve(provider),
         {:ok, out} <- Backend.run_provider(boot, "url", Map.put(Config.to_env(p), "WB_IMAGE", Image.ref())),
         url when is_binary(url) <- (out |> String.split("\n") |> Enum.map(&String.trim/1) |> Enum.find(&String.starts_with?(&1, "http"))),
         {:ok, 200, body} <- http_get(url <> "/health", nil) do
      if String.contains?(body, "ok"),
        do: ok("cloud runtime healthy — #{url}/health → 200", %{state: "healthy", url: url, provider: provider}),
        else: err("cloud runtime reachable but unhealthy at #{url}", %{state: "unhealthy", url: url})
    else
      {:ok, status, _} -> err("cloud runtime returned HTTP #{status}", %{state: "unhealthy", http: status})
      _ -> err("could not verify cloud runtime (provider #{provider}) — is it deployed? `wb deploy apply`", %{state: "unreachable"})
    end
  end

  @doc "Render a tagged result for the CLI — `--json` → machine map, else the message."
  def render({tag, msg, data}, json?) do
    if json?,
      do: Jason.encode!(Map.merge(%{ok: tag == :ok, message: msg}, data)),
      else: if(tag == :ok, do: msg, else: "deploy error: #{msg}")
  end

  @doc "Did a verb fail? (drives the CLI exit code.)"
  def failed?({:error, _, _}), do: true
  def failed?(_), do: false

  # ---- internals -------------------------------------------------------------
  defp ok(msg, data), do: {:ok, msg, data}
  defp err(msg, data), do: {:error, msg, data}

  defp await_discovery(timeout) do
    deadline = System.monotonic_time(:millisecond) + timeout
    do_await(deadline)
  end

  defp do_await(deadline) do
    case Krunvm.discovery() do
      {:ok, d} ->
        {:ok, d}

      _ ->
        if System.monotonic_time(:millisecond) < deadline do
          Process.sleep(500)
          do_await(deadline)
        else
          {:error, :timeout}
        end
    end
  end

  defp http_get(url, token) do
    :inets.start()
    :ssl.start()
    headers = if token, do: [{~c"authorization", String.to_charlist("Bearer " <> token)}], else: []

    case :httpc.request(:get, {String.to_charlist(url), headers}, [timeout: 5_000], body_format: :binary) do
      {:ok, {{_, status, _}, _, body}} -> {:ok, status, body}
      {:error, reason} -> {:error, reason}
    end
  end

  # A free localhost port for the host→guest map (the guest always binds 4000).
  defp free_host_port do
    {:ok, s} = :gen_tcp.listen(0, [:binary, ip: {127, 0, 0, 1}])
    {:ok, p} = :inet.port(s)
    :gen_tcp.close(s)
    p
  end
end

File: cli/src/deploy.zig
//! deploy — stand up a runtime for a workbook, local or cloud. The config is a single `<work-deploy>`
//! HTML element (no JSON, on-canon). `init` scaffolds it; `validate` runs the coherence rules (the
//! same lockouts the Elixir kit had); apply/status/verify/down route by engine-place to a backend.
//! Config parse + validation are pure (testable); secrets never live in the file — they come from ENV.
const std = @import("std");
const Io = std.Io;
const fs = @import("fs.zig");
const log = @import("log.zig");

const places = [_][]const u8{ "local", "cloud" };
const tenancies = [_][]const u8{ "single", "multi" };
const storages = [_][]const u8{ "local-fs", "s3" };
const databases = [_][]const u8{ "sqlite", "postgres" };
const auths = [_][]const u8{ "trusted", "betterauth", "clerk", "oidc" };

pub fn scaffold(alloc: std.mem.Allocator, place: []const u8) ![]const u8 {
    const cloud = std.mem.eql(u8, place, "cloud");
    return std.fmt.allocPrint(alloc,
        \\<!doctype html>
        \\<html lang="en"><head><meta charset="utf-8"><title>Deployment</title></head><body>
        \\<!-- secrets (WB_DATABASE_URL / WB_S3_*) come from your deploy ENV, never this file -->
        \\<work-deploy
        \\  engine-place="{s}"
        \\  tenancy-mode="single"
        \\  storage="{s}"
        \\  database="{s}"
        \\  auth="trusted"{s}>
        \\</work-deploy>
        \\</body></html>
        \\
    , .{
        place,
        if (cloud) "s3" else "local-fs",
        if (cloud) "postgres" else "sqlite",
        if (cloud) "\n  provider=\"fly\"\n  app=\"my-workbook\"\n  region=\"iad\"" else "",
    });
}

pub fn attr(html: []const u8, name: []const u8, default: []const u8) []const u8 {
    const wd = std.mem.indexOf(u8, html, "<work-deploy") orelse return default;
    const end = std.mem.indexOfScalarPos(u8, html, wd, '>') orelse html.len;
    const region = html[wd..@min(end, html.len)];
    var pat: [64]u8 = undefined;
    const p = std.fmt.bufPrint(&pat, "{s}=\"", .{name}) catch return default;
    const at = std.mem.indexOf(u8, region, p) orelse return default;
    const vs = at + p.len;
    const ve = std.mem.indexOfScalarPos(u8, region, vs, '"') orelse return default;
    return region[vs..ve];
}

/// Validate a parsed config → list of issue strings ([] means coherent).
pub fn validate(alloc: std.mem.Allocator, html: []const u8) ![]const []const u8 {
    var issues: std.ArrayList([]const u8) = .empty;
    const place = attr(html, "engine-place", "local");
    const tenancy = attr(html, "tenancy-mode", "single");
    const storage = attr(html, "storage", "local-fs");
    const database = attr(html, "database", "sqlite");
    const auth = attr(html, "auth", "trusted");

    try enumCheck(alloc, &issues, "engine-place", place, &places);
    try enumCheck(alloc, &issues, "tenancy-mode", tenancy, &tenancies);
    try enumCheck(alloc, &issues, "storage", storage, &storages);
    try enumCheck(alloc, &issues, "database", database, &databases);
    try enumCheck(alloc, &issues, "auth", auth, &auths);

    if (eql(tenancy, "multi") and eql(database, "sqlite"))
        try issues.append(alloc, "tenancy-mode: multi needs database: postgres (sqlite can't isolate tenants)");
    if (eql(tenancy, "multi") and eql(auth, "trusted"))
        try issues.append(alloc, "tenancy-mode: multi needs real auth (betterauth|clerk|oidc) — trusted has no identity");
    if (eql(place, "cloud") and eql(auth, "trusted"))
        try issues.append(alloc, "engine-place: cloud + auth: trusted is an OPEN control plane — set WB_PUBLIC_BEARER in your deploy ENV, or use real auth");
    if (eql(storage, "s3") and (attr(html, "storage-bucket", "").len == 0 or attr(html, "storage-endpoint", "").len == 0))
        try issues.append(alloc, "storage: s3 needs storage-bucket + storage-endpoint");
    if (eql(database, "postgres"))
        try issues.append(alloc, "database: postgres needs WB_DATABASE_URL in your deploy ENV");

    return issues.toOwnedSlice(alloc);
}

// ── verbs ───────────────────────────────────────────────────────────────────────────────────
pub fn init(io: Io, alloc: std.mem.Allocator, place_in: []const u8, dir: []const u8, force: bool) !u8 {
    const place = if (isPlace(place_in)) place_in else "local";
    const path = try std.fs.path.join(alloc, &.{ dir, "deployment.html" });
    log.prompt(try std.fmt.allocPrint(alloc, "work deploy init {s} {s}", .{ place, dir }));

    if (!force) {
        if (fs.readFile(io, alloc, path)) |_| {
            log.warn("deployment.html already exists — pass --force to overwrite");
            return 1;
        } else |_| {}
    }
    try fs.writeFile(io, path, try scaffold(alloc, place));
    log.ok(try std.fmt.allocPrint(alloc, "wrote deployment.html \u{b7} engine-place={s}", .{place}));
    log.step("edit it, then `work deploy validate` \u{2192} `work deploy apply`");
    return 0;
}

pub fn validateVerb(io: Io, alloc: std.mem.Allocator, file: []const u8) !u8 {
    log.prompt(try std.fmt.allocPrint(alloc, "work deploy validate {s}", .{file}));
    const html = fs.readFile(io, alloc, file) catch {
        log.err("no deployment.html — run `work deploy init` first");
        return 1;
    };
    if (std.mem.indexOf(u8, html, "<work-deploy") == null) {
        log.err("no <work-deploy> element in the config");
        return 1;
    }
    const issues = try validate(alloc, html);
    if (issues.len == 0) {
        log.ok(try std.fmt.allocPrint(alloc, "config is coherent \u{b7} engine-place={s}", .{attr(html, "engine-place", "local")}));
        return 0;
    }
    log.err(try std.fmt.allocPrint(alloc, "{d} issue(s)", .{issues.len}));
    for (issues) |i| log.step(i);
    return 1;
}

pub fn apply(io: Io, alloc: std.mem.Allocator, file: []const u8) !u8 {
    log.prompt(try std.fmt.allocPrint(alloc, "work deploy apply {s}", .{file}));
    const html = fs.readFile(io, alloc, file) catch {
        log.err("no deployment.html — run `work deploy init` first");
        return 1;
    };
    const issues = try validate(alloc, html);
    if (issues.len > 0) {
        log.err("config invalid — fix it first (`work deploy validate`)");
        for (issues) |i| log.step(i);
        return 1;
    }
    const place = attr(html, "engine-place", "local");
    if (eql(place, "local")) {
        log.ok("local target — runs the one OCI image in a krunvm/container");
        log.step("image build + boot lands with the nexus image recipe");
    } else {
        log.ok(try std.fmt.allocPrint(alloc, "cloud target: {s} \u{b7} {s}", .{ attr(html, "provider", "fly"), attr(html, "app", "(no app)") }));
        log.step("cloud apply provisions via the control plane (`work login`)");
    }
    return 0;
}

fn enumCheck(alloc: std.mem.Allocator, issues: *std.ArrayList([]const u8), key: []const u8, value: []const u8, allowed: []const []const u8) !void {
    for (allowed) |a| if (eql(a, value)) return;
    try issues.append(alloc, try std.fmt.allocPrint(alloc, "{s}: \"{s}\" is not valid", .{ key, value }));
}

fn eql(a: []const u8, b: []const u8) bool {
    return std.mem.eql(u8, a, b);
}
fn isPlace(p: []const u8) bool {
    for (places) |x| if (eql(x, p)) return true;
    return false;
}

test "scaffold → validate: coherent local, flagged cloud+multi+trusted+s3" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();

    const local = try scaffold(a, "local");
    try std.testing.expectEqual(@as(usize, 0), (try validate(a, local)).len);

    const bad = "<work-deploy engine-place=\"cloud\" tenancy-mode=\"multi\" storage=\"s3\" database=\"postgres\" auth=\"trusted\">";
    const issues = try validate(a, bad);
    try std.testing.expect(issues.len >= 3);
}
